### PR TITLE
[WIP] Upgrade solidity-coverage to 0.7.0

### PR DESCRIPTION
**Pull request by cgewecke**
(Initial discussion about this is in #1918.  Opening as WIP because `0.7.0` is in beta and I'd like some feedback on whether the changes in this PR make sense - it's possible they should be broken into separate submissions).

`solidity-coverage` has a new design - it now runs as a plugin and can use any version of ganache. It launches the client itself because it needs to hook into the VM in memory in order to track code execution during the tests. 

As a result installation is a bit different - ganache options which are passed as flags to a stand-alone client need to be formatted as [ganache-core options](https://github.com/trufflesuite/ganache-core#options) and listed in the coverage config file under the `providerOptions` key.

Additionally, SC no longer instruments Solidity dependencies or modifies contract ABIs and many of its quirks have been resolved. Abnormal gas consumption is also less of an issue (all tests complete without problems here atm).  (There's more detail about this work at [solidity-coverage 372](https://github.com/sc-forks/solidity-coverage/pull/372). Info about installation & new options can be found [here](https://github.com/sc-forks/solidity-coverage/tree/truffle-plugin#install))

**PR**:
+ Extracts account definitions from `test.sh` to their own file because this data needs to be accessed in several places (test.sh, truffle-config.js, .solcover.js)
+ Adds JS helpers to load accounts info in the formats required by truffle-config & solcover.js
+ Adds a JS helper to run `oz-gsn` relay hub so the coverage plugin can execute that step itself after it launches ganache.
+ Sets solidity-coverage version to `0.7.0-beta.2` 
+ Removes ganache-cli-coverage package.
+ Removes all coverage logic in `test.sh` in favor of a small change to `coverage.sh`
+ Removes coverage network in truffle-config.js and loads its GSN account info using the JS helper described above. 

(Leaving a few more comments in the diff where stuff might seem weird)

[Edit 1 - CI isn't running...not sure what's wrong. I have enabled it for my fork and the PR [runs ok there](https://circleci.com/workflow-run/274c3316-719b-42d3-bc58-2311f894c597)]

### CODE REVIEW

File: scripts/shared.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+        reject(err);
+      }
+    });
+
+    gsnProcess.stdout.on('data', data => {
+      log(`oz-gsn stdout: ${data}`);
+      if (data.includes('RelayHub deployed at')) resolve();
+    });
+
+    gsnProcess.stderr.on('data', data => {
```
Comment (cgewecke) [inline scripts/shared.js]: Not sure if child.exec is how you'd actually like to do this... @nventuro mentioned in #1918 that GSN has it's own JS API. 

In any case - there might an issue here with how oz-gsn logs it's data - it's mostly coming out of `stderr` so am listening on & resolving from both streams. Have asked a question about that in [oz-gsn-helpers 21](https://github.com/OpenZeppelin/openzeppelin-gsn-helpers/issues/21)

Comment (cgewecke) [reply scripts/shared.js]: Per [comment](https://github.com/OpenZeppelin/openzeppelin-gsn-helpers/issues/21#issuecomment-536732477) at issue ref'd above, if this exec approach is kept the stdout listener should be removed.

File: scripts/coverage.sh [hunk 2]
```diff
@@ -6,7 +6,7 @@ log() {
   echo "$*" >&2
 }
 
-SOLIDITY_COVERAGE=true scripts/test.sh || log "Test run failed"
+npx truffle run coverage --network development
```
Comment (nventuro) [inline scripts/coverage.sh]: Is this the syntax for running a truffle plugin?

File: .solcover.js [hunk 3]
```diff
@@ -1,9 +1,13 @@
...
-        'lifecycle/Migrations.sol',
-        'mocks'
-    ]
+  skipFiles: [
+    'lifecycle/Migrations.sol',
+    'mocks'
+  ],
+  port: 8545,
+  client: ganache,
+  onServerReady: shared.setupGSNRelayHub,
```
Comment (nventuro) [inline .solcover.js]: Is there documentation for these options? I'm assuming SC will run `client` on `port` with the options provided, and then call `onServerReady`, but does this happen on a separate process?

File: scripts/coverage.sh (hunk 2 above)
Comment (cgewecke) [reply scripts/coverage.sh]: Yes. There's more at the [truffle docs for plugins](https://www.trufflesuite.com/docs/truffle/getting-started/writing-external-scripts#third-party-plugin-commands).

File: .solcover.js (hunk 3 above)
Comment (cgewecke) [reply .solcover.js]: @nventuro 

Yes, there are [options docs here](https://github.com/sc-forks/solidity-coverage/tree/beta#config-options).

>  I'm assuming SC will run client on port with the options provided, and then call onServerReady, but does this happen on a separate process?

The ganache server is run the same process as truffle. In this PR, `onServerReady` executes a [helper method](https://github.com/OpenZeppelin/openzeppelin-contracts/pull/1923/files#diff-f76afefa4e67ba58e941de1756c6b919R68) which uses child_process.exec to launch `oz-gsn` separately.

**Comment by nventuro** [comment]
Hello @cgewecke, thank you very much for taking the time to tackle this! I'm sorry my request required you to refactor SC this much :sweat_smile: 

Regarding this PR, it seems like there's a lot of accidental complexity in `shared.js` due to the need to read the accounts from two different places (and environments) because of how SC is configured.

I don't know the details behind how truffle plugins work, so I'm not sure if there are any limitations or proposed design patterns, but it feels like there should be a way to make this simpler for the user. My naive suggestion is as follows:
 - keep a `coverage` command that instruments, runs all tests, and then measures coverage
 - add standalone commands for these tree steps (e.g. `coverage setup`, `coverage run`, `coverage measure`), so that more granular control is provided for complex setups (like this one)

This way, we would have a a single testing setup script (in Bash most likely), shared by regular tests and coverage runs (with some `if`s based on arguments or envvars).

Do these suggestions make sense at all, or am I misunderstanding something about SC or this PR?

**Comment by cgewecke** [comment]
@nventuro No worries at all, thanks so much for taking the time to look at this. 

Could I ask which stages of the `test.sh` script your proposed sub-commands map to? 

> there's a lot of accidental complexity in shared.js due to the need to read the accounts from two different places

I agree this is adding a bunch of helper code. One defense might be that there *already is* multi-context reference to the account info - for example in [truffle-config.js](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/truffle-config.js#L22-L23) where the addresses are hard-coded. 

There are also hardcoded [references to the RelayHub address](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/truffle-config.js#L22-L23) in the GSN tests which might fit into this 'shared address' category.  

I will need to think about what you've suggested more - my gut feeling is that supporting this alternate sub-command sequence would add quite a bit more complexity to solidity-coverage and that it's slightly harder to explain than the [idea of workflow hooks](https://github.com/sc-forks/solidity-coverage/blob/beta/docs/advanced.md#workflow-hooks) (which is the current strategy for managing advanced use contexts.) But am not completely fixed in that view fwiw.

As discussed in #1918 - a core goal of the `0.7.0` redesign is make the pieces of SC available so that any tooling can consume them as it sees fit. That work should be finalized in ~1 month - maybe we should re-visit this topic then because you might be able to build exactly what you want.)

**Comment by cgewecke** [comment]
@nventuro Have thought about this a bit more....
 
+ Agree the `readline` helpers aren't the best solution.
+ However, I think `test.sh` might benefit from a redesign. Routing all command flow through a if/elses gated by env variables in a shell script has disadvantages.... 
+ ...for example: 🙂 To preserve the `test.sh` flow you've proposed that SC expose its internals as three separate shell commands, each of which would need to serialize state and save it to the filesystem so subsequent commands can pick up where the last left off. None of these commands is useful on its own - it's kind of unusual. 

An alternative is to modify the tests setup here so its components can be used in different ways (including optionally). For example, you could:

+ Place hardcoded account variables and network values in a json file analogous to a `genesis.json` so all data is in a single place / standard format. 
+ Use a cli utility like [json](https://www.npmjs.com/package/json) to manage network info in shell scripts. JS files can `require` it.
+ Make the shell tasks more discrete and compose larger tasks as npm commands: 

   **Example**:
   ```json
   "startClient": "./scripts/startClient.sh",
   "stopClient": "./scripts/stopClient.sh",
   "test": "npm run startClient && npm run test:units && npm run stopClient",
   "test:units": "npx truffle test --network development",
   "codecov" : "curl -s https://codecov.io/bash | bash -s -- -C $CIRCLE_SHA1",
   "coverage": "npx truffle run coverage --network development",
  ```
+ provide a javascript GSN relay hub launcher by default, so external tooling can integrate it as necessary.

One benefit of the above is that you'd provide analytics tools good entry points to interact with the Zeppelin contracts, whatever their design. 

SC aside, this repo is the first stop for everyone building anything.  It's every tool's frontline E2E target because it's so well-written, well-tested, well-maintained, well-designed, etc. If something built works here, it works. 

Essentially I'm saying there's a case for organizing the repo itself *as if it provided an API*.

**Comment by nventuro** [comment]
>However, I think test.sh might benefit from a redesign

Yes, this is very clear from this conversation. It is also the case that part of the history of `test.sh` was guided by the old coverage API :) So we will be revaluating this and other tools to come up with a more cohesive solution.

Unless this is now blocking your e2e tests for SC, what I suggest we do is pause integration of the beta coverage branch into this repo until the API is finalized and we have a clearer picture of how we'll move forward (we have multiple projects across a number of repos and would like to somehow unify this setup). 

Thank you very much for your work on this: our discussions have been extremely useful for us to reflect on these topics. I find your point about thinking of the repo as an E2E target particularly intriguing.

**Comment by cgewecke** [comment]
> Unless this is now blocking your e2e tests for SC, what I suggest we do is pause integration of the beta coverage branch into this repo until the API is finalized and we have a clearer picture of how we'll move forward (we have multiple projects across a number of repos and would like to somehow unify this setup).

Sounds good! I am running E2E on our zeppelin fork and am very happy with that. AFAIK this PR can be deferred indefinitely - your solidity-coverage versions and/or `0.6.x` should continue to work until mainnet forks again. 

I leave it to you whether to keep this open or closed. Thanks so much @nventuro!

**Comment by villesundell** [comment]
Hello everyone! 0.7.0 was released 8 days ago (https://github.com/ethereum/solidity/releases/tag/v0.7.0), so support for it would be greatly appreciated :)

**Review by mkushka — APPROVED** [review]
Looks fine but the suggestions were made about 1 year ago...(

**Comment by frangio** [comment]
Fixed in https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2291.

**Pull request closed without merging**